File: src/main/java/dage/visualhelditems/CobblemonHeldItems.java
package dage.visualhelditems;

import com.cobblemon.mod.common.api.Priority;
import com.cobblemon.mod.common.api.events.CobblemonEvents;
import com.cobblemon.mod.common.entity.pokemon.PokemonEntity;
import net.fabricmc.api.ModInitializer;
import net.fabricmc.fabric.api.networking.v1.EntityTrackingEvents;
import net.fabricmc.fabric.api.networking.v1.PacketByteBufs;
import net.fabricmc.fabric.api.networking.v1.ServerPlayNetworking;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.network.PacketByteBuf;
import net.minecraft.registry.RegistryKeys;
import net.minecraft.registry.tag.TagKey;
import net.minecraft.server.MinecraftServer;
import net.minecraft.server.network.ServerPlayerEntity;
import net.minecraft.util.Identifier;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.HashMap;
import java.util.HashSet;
import java.util.UUID;


public class CobblemonHeldItems implements ModInitializer {
	public static final String MOD_ID = "cobblemon_helditems";
	public static final Logger LOGGER = LoggerFactory.getLogger(MOD_ID);
	/**
	 * A list of pokemon entities currently being tracked on the server.
	 */
	private static final HashMap<UUID, HashSet<ServerPlayerEntity>> cachedTrackedPokemon = new HashMap<>();

	public static final TagKey<Item> HIDDEN_ITEMS = TagKey.of(RegistryKeys.ITEM, new Identifier("cobblemon_helditems", "hidden_items"));
	public static final Identifier ENTITY_START_TRACKING = new Identifier(MOD_ID, "pokemon_item_start");
	public static final Identifier ENTITY_STOP_TRACKING = new Identifier(MOD_ID, "pokemon_item_stop");
	public static final Identifier HELD_ITEM_UPDATED = new Identifier(MOD_ID, "pokemon_item_updated");

	@Override
	public void onInitialize() {
		CobblemonEvents.HELD_ITEM_POST.subscribe(Priority.NORMAL, post -> {
			PokemonEntity pokemonEntity = post.getPokemon().getEntity();
			//TODO find a better way to send updates from pokemon to player trackers
			if (pokemonEntity != null) {
				UUID pokemonId = pokemonEntity.getUuid();
				if (cachedTrackedPokemon.containsKey(pokemonId)) {
					for (ServerPlayerEntity player : cachedTrackedPokemon.get(pokemonId)){
						MinecraftServer server = player.getServer();
						ItemStack heldItem = post.getReceived();
						if(heldItem.isIn(HIDDEN_ITEMS)) heldItem = ItemStack.EMPTY;

						PacketByteBuf data = PacketByteBufs.create();
						data.writeUuid(pokemonId);
						data.writeItemStack(heldItem);

						// Send a packet to the client
						ServerPlayerEntity playerEntity = server.getPlayerManager().getPlayer(player.getUuid());
						server.execute(() -> {
							ServerPlayNetworking.send(playerEntity, HELD_ITEM_UPDATED, data);
						});
					}
				}
			}
			return null;
		});

		EntityTrackingEvents.START_TRACKING.register( (trackedEntity, player) -> {
			if (trackedEntity.getClass() == PokemonEntity.class) {
				if ( ((PokemonEntity) trackedEntity).getPokemon().getOwnerUUID() == player.getUuid() ) return;//exit early if entity can be rendered client side.

				PokemonEntity pokemonEntity = (PokemonEntity)trackedEntity;
				UUID pokemonId = pokemonEntity.getUuid();
				ItemStack heldItem = pokemonEntity.getPokemon().heldItem();
				if(heldItem.isIn(HIDDEN_ITEMS)) heldItem = ItemStack.EMPTY;

				MinecraftServer server = player.getServer();

				PacketByteBuf data = PacketByteBufs.create();
				data.writeUuid(pokemonId);
				data.writeItemStack(heldItem);

				// Send a packet to the client
				ServerPlayerEntity playerEntity = server.getPlayerManager().getPlayer(player.getUuid());
				server.execute(() -> {
					ServerPlayNetworking.send(playerEntity, ENTITY_START_TRACKING, data);

					if ( cachedTrackedPokemon.containsKey(pokemonId) ) cachedTrackedPokemon.get(pokemonId).add(playerEntity);
					else {
						cachedTrackedPokemon.putIfAbsent(pokemonId, new HashSet<>() );
						cachedTrackedPokemon.get(pokemonId).add(playerEntity);
					}
				});
			}
		});

		EntityTrackingEvents.STOP_TRACKING.register((trackedEntity, player) -> {
			if (trackedEntity.getClass() == PokemonEntity.class) {
				if ( ((PokemonEntity) trackedEntity).getPokemon().getOwnerUUID() == player.getUuid() ) return;//exit early if entity can be rendered client side.

				PokemonEntity pokemonEntity = (PokemonEntity)trackedEntity;
				UUID pokemonId = pokemonEntity.getUuid();
				MinecraftServer server = player.getServer();

				PacketByteBuf data = PacketByteBufs.create();
				data.writeUuid(pokemonId);

				// Send a packet to the client
				ServerPlayerEntity playerEntity = server.getPlayerManager().getPlayer(player.getUuid());
				server.execute(() -> {
					ServerPlayNetworking.send(playerEntity, ENTITY_STOP_TRACKING, data);
					cachedTrackedPokemon.get(pokemonId).remove(playerEntity);
					if (cachedTrackedPokemon.get(pokemonId).isEmpty()) cachedTrackedPokemon.remove(pokemonId);
				});
			}
		});
	}
}

File: common/src/main/kotlin/com/dage/visualhelditems/CobblemonHeldItems.kt
package com.dage.visualhelditems

import com.cobblemon.mod.common.api.Priority
import com.cobblemon.mod.common.api.events.CobblemonEvents.HELD_ITEM_POST
import com.cobblemon.mod.common.api.events.pokemon.HeldItemEvent
import com.cobblemon.mod.common.client.CobblemonClient.storage
import com.cobblemon.mod.common.client.storage.ClientPC
import com.cobblemon.mod.common.entity.pokemon.PokemonEntity
import net.minecraft.client.MinecraftClient
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraft.registry.RegistryKeys
import net.minecraft.registry.tag.TagKey
import net.minecraft.util.Identifier
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.util.*

object CobblemonHeldItems {
    const val MOD_ID = "cobblemon_helditems"

    val LOGGER: Logger = LoggerFactory.getLogger(MOD_ID)
    /**
     * A list of pokemon entities currently being tracked on the server.
     */
    val cachedTrackedPokemon: HashMap<UUID, ItemStack> = HashMap()

    val WEARABLE_EYE_ITEMS: TagKey<Item> = TagKey.of(RegistryKeys.ITEM, Identifier("cobblemon_helditems", "wearable_eye_items"))
    val WEARABLE_HAT_ITEMS: TagKey<Item> = TagKey.of(RegistryKeys.ITEM, Identifier("cobblemon_helditems", "wearable_hat_items"))
    val HIDDEN_ITEMS: TagKey<Item> = TagKey.of(RegistryKeys.ITEM, Identifier("cobblemon_helditems", "hidden_items"))

    fun updateShownItem(pokemonEntity :PokemonEntity, item: ItemStack) {
        var shownItem = ItemStack.EMPTY;
        if (!item.isIn(HIDDEN_ITEMS)) shownItem = item;

        if (pokemonEntity is ShownItemTracker) {
            (pokemonEntity as ShownItemTracker).shownItem = shownItem
        }
    }


    fun getHeldItem(pokemonEntity: PokemonEntity): ItemStack? {
        val pokemonID : UUID = pokemonEntity.uuid;
        //TODO Make the client storage search not so insanely bad..
        if (pokemonEntity.ownerUuid != MinecraftClient.getInstance().player?.uuid){
            if (pokemonEntity is ShownItemTracker) {
                val i = pokemonEntity.shownItem
                if (!i.isEmpty) return i
            }
        } else {
            val storage = storage
            val party = storage.myParty
            val pcs: Collection<ClientPC> = storage.pcStores.values
            //See if the pokemon that is being rendered is part of client users party
            for (p in party) {
                if (p == null) continue
                val partyEntity = p.entity ?: continue
                if (partyEntity.uuid === pokemonID) return p.heldItem()
            }
            //If not then check PC **this is from pokemon roaming around from the pasture block**
            //AKA triple for-loop nightmare
            for (pc in pcs) {
                for (box in pc.boxes) {
                    for (p in box.slots) {
                        if (p == null) continue
                        val partyEntity = p.entity ?: continue
                        if (partyEntity.uuid === pokemonID) return p.heldItem()
                    }
                }
            }
        }
        return ItemStack.EMPTY
    }

    fun initialize() {
        HELD_ITEM_POST.subscribe(Priority.NORMAL) { post: HeldItemEvent.Post ->
            val pokemonEntity: PokemonEntity? = post.pokemon.entity

            if (pokemonEntity != null) {
                updateShownItem(pokemonEntity, post.received)
            }
        }
    }
}